### ISSUE TITLE
AwsJsonProtocolErrorUnmarshaller does not handle GZIP responses

### ISSUE THREAD

**Issue by cristianocca** [aws/aws-sdk-java-v2#4710]
### Describe the bug

We are using an auto generated SDK client from smithy to java code generation.

The [following](https://github.com/aws/aws-sdk-java-v2/blob/master/core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/unmarshall/AwsJsonProtocolErrorUnmarshaller.java#L66) code does not handle GZIP responses, and always parses as an empty json response.

I'm not sure whether this is a bug or simply not supported, but regular responses are deserialized correctly.

### Expected Behavior

Error responses should be deserialized correctly just like regular responses.

### Current Behavior

Error responses are not properly deserialized.

Code that fails: https://github.com/aws/aws-sdk-java-v2/blob/master/core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/unmarshall/AwsJsonProtocolErrorUnmarshaller.java#L66 and https://github.com/aws/aws-sdk-java-v2/blob/master/core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/JsonContent.java#L76

### Reproduction Steps

Send an error response from the target server and observe how the sdk client fails to parse it.

### AWS Java SDK version used

v2

### JDK version used

JDK 17

### Operating System and version

Linux

**Comment by debora-ito** [comment]
Hi @cristianocca what do you mean by GZIP responses, exactly? Can you give an example?

**Comment by cristianocca** [comment]
The server returns the response in GZIP encoding. The SDK client handles gzipped responses just fine when it is a success response ([this](https://github.com/aws/aws-sdk-java-v2/blob/cb17af86d25fa4feb885e3dd7c4aa9ec3f9d5888/core/sdk-core/src/main/java/software/amazon/awssdk/core/http/Crc32Validation.java#L85) class is used), but fails to decompress the content when running through the error handling case (the Crc32 validation is never used).

Seems like [this](https://github.com/aws/aws-sdk-java-v2/blob/cb17af86d25fa4feb885e3dd7c4aa9ec3f9d5888/core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/CombinedResponseHandler.java#L81) code does not set up the error handler to also include the decompression class.

Sadly, the only work around is to go to the server and disable gzip for any response that is not HTTP 200.

**Comment by debora-ito** [comment]
@cristianocca Instead of pointing to the SDK codebase, can you provide a code sample and a response syntax to illustrate the issue you're running into? I'm looking for a SSCCE code sample - http://sscce.org/

I see you're using the SDK to generate a client for your service, but I need more details to fully understand the issue.

Please also provide:
- The version of the SDK you're using (like 2.21.3)
- The full stacktrace of the error
- Which tool are you using to generate the Java client from the Smithy model?

**Comment by cristianocca** [comment]
Hi @debora-ito , I cannot really share code examples as this is internal.  I did open a ticket internally (amazon) which I think will be better to share information. I was a bit confused as to where to report the bug since this package is both public and internal.

Regarding the requested information.
- SDK: 2.21.16 (public version)
- Stacktrace of the error: there's no stack trace as there's really not an error/exception. As I linked above, [this](https://github.com/aws/aws-sdk-java-v2/blob/master/core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/JsonContent.java#L76) fails to parse the json response from the server because the actual response body is gzipped and it was never decompressed. Below is a sample of what's logged (redacted).
- Using Smithy2C2JBuild and AwsJavaSdk-Codegen-BuildTools to generate the client code from a smithy model.

Sample log:
```
software.amazon.awssdk.thirdparty.jackson.core.JsonParseException: Illegal character ((CTRL-CHAR, code 31)): only regular white space (\r, \n, \t) is allowed between tokens
 at [Source: (byte[])"\u001F�\u0008....<more bytes>....
```
